### ISSUE TITLE
[EPIC] Support signal events

### ISSUE THREAD

**Issue by korthout** [camunda/zeebe#10777]
## Introduction

In process automation, sometimes there is a need to broadcast a signal to one or multiple waiting process instances or processes.

> A BPMN Signal is similar to a signal flare that [is] shot into the sky for anyone who might be interested to notice and then react. Thus, there is a source of the Signal, but no specific intended target. 
>-[BPMN 2.0.2 - 10.5.4 Intermediate Event](https://www.omg.org/spec/BPMN/2.0.2/PDF#10.5.4%20Intermediate%20Event)

>Signals [...] are typically used for broadcast communication within and across Processes, across Pools, and between Process diagrams.
>-[BPMN 2.0.2 - 10.5.1 Concepts](https://www.omg.org/spec/BPMN/2.0.2/PDF#10.5.1%20Concepts)

As a user, I can:
- deploy all Signal Events to the Engine (Zeebe).
- be sure that the Engine uses signal events correctly, e.g.
  - triggers signals when the token arrives at throwing signal events
  - and broadcast to all catching signal events (including Start Event)
- broadcast a signal via the gRPC API

## Types of Signal Events

<img width="692" alt="Table 10.93 - Types of Signal Events and their Markers (cropped)" title="Table 10.93 - Types of Signal Events and their Markers (cropped)" src="https://user-images.githubusercontent.com/3511026/196949501-8ce4b8da-ae64-440a-a5ba-aaaffad44e84.png">

<details>
<summary><h3>Specification of Signal Event Types</h3></summary>
<p>

The specification defines the following types of Signal Events and their behaviors:

#### Top-Level Process Start Events
<img width="687" alt="Table 10.84 – Top-Level Process Start Event Types (cropped)" title="Table 10.84 – Top-Level Process Start Event Types (cropped)" src="https://user-images.githubusercontent.com/3511026/196949757-c9397587-2968-45d9-8127-860210621a0d.png">

#### Event Sub-Process Start Events
<img width="694" alt="Table 10.86 – Event Sub-Process Start Event Types (cropped)" title="Table 10.86 – Event Sub-Process Start Event Types (cropped)" src="https://user-images.githubusercontent.com/3511026/196951294-62481cd2-2560-4c11-92c1-3fe44d70a043.png">

#### Intermediate Events
<img width="694" alt="Table 10.89 – Intermediate Event Types in Normal Flow (cropped)" title="Table 10.89 – Intermediate Event Types in Normal Flow (cropped)" src="https://user-images.githubusercontent.com/3511026/196950488-c7e45f60-03a2-4893-82ab-dd25afcf9d0f.png">

#### Boundary Events
<img width="691" alt="Table 10.90 – Intermediate Event Types Attached to an Activity Boundary (cropped)" title="Table 10.90 – Intermediate Event Types Attached to an Activity Boundary (cropped)" src="https://user-images.githubusercontent.com/3511026/196951138-171d1b3f-66aa-4a45-96fc-06569430fed6.png">

#### End Events
<img width="696" alt="Table 10.88 – End Event Types (cropped)" title="Table 10.88 – End Event Types (cropped)" src="https://user-images.githubusercontent.com/3511026/196951075-a599054f-5860-4720-8300-813380e5e970.png">

</p>
</details>

## Concept

Signals are broadcasted to both processes (received by top-level signal start events) and process instances (received by all other signal catch events).

### Top-level start events
A signal is broadcasted to all processes, so it can be caught by a signal top-level start event, starting a new instance of that process when received. 

In Zeebe, this should be straightforward to implement. The process instances can be created on the same partition that the signal is broadcasted from (i.e. the partition where the signal throw event is activated; or the partition that has received the signal broadcast request from the gateway). In other words, the signal top-level start events can catch the signal on the same partition. We have no need for inter-partition communication for this part of the signal logic.

> **Note**
> We only have to create an instance of the latest version of each of the deployed processes known by that partition at the moment that the signal is received.

### Other catch events
A signal is also broadcasted to all process instances, so it can be caught by any of the other signal catch events when received. Those process instances that are awaiting the signal can continue their execution once they receive the signal.

In Zeebe, a process instance exists only on one partition. So, a signal must be broadcasted to all the partitions that have process instances. We can simply broadcast it to all partitions. Each partition receives the signal and then continues all the process instances that are awaiting that signal. So we need inter-partition communication for this part of the signal logic.

### Signal logic overview
Here's a diagram describing both parts of the signal logic and the inter-partition communication.

![Screen Shot 2022-10-20 at 15 12 40](https://user-images.githubusercontent.com/3511026/196958193-747e7e84-3cea-4123-b507-abfadee4059b.png)

<details><summary><h3>Alternative concept discussion</h3></summary>
<p>

The concept described above leads to two main performance challenges. Let's discuss these individually.

1. Each partition may have many process instances awaiting a signal. A broadcasted signal could produce lots of work for the engine on each partition. The amount of work is determined by the number of process instances awaiting that specific signal.
2. Each broadcasted signal requires communication between the partitions. This communication traffic increases linearly with the number of partitions.

There's not much we can do to reduce the impact of the first challenge because it depends directly on the number of process instances awaiting that specific signal. It's up to the user to decide how they design their processes and how many process instances are awaiting a signal. We should inform users about the performance impact that comes with the usage of signals. Over-usage must be avoided.

> **Note**
> Due to a technical limitation, there's an actual limit to the number of catch events that a signal can be caught by. At the time of writing, a processed command can only produce a limited number of follow-up records. This limitation is defined by `maxMessageSize`.

The impact of the second challenge can be reduced by communicating only with those partitions that have process instances awaiting the specific signal. We don't have to broadcast the signal to partitions that have no process instances. Likewise, we don't have to broadcast the signal to partitions that have process instances, but none of them use signals. Etc.

In order for a partition to know whether another partition has process instances awaiting a specific signal, some form of bookkeeping needs to take place on both partitions. This bookkeeping means inter-partition communication in the opposite direction.

![Screen Shot 2022-10-20 at 15 20 22](https://user-images.githubusercontent.com/3511026/196959939-f53ab450-e295-46b9-a236-45d6baf9656a.png)

When looking at it in a diagram, it quickly becomes clear that this defeats the entire purpose. Each signal now requires multiple communications. This only improves the situation when a few catch events receive a signal. Generally, we expect that each signal is received by many catch events. 

To make matters worse, a signal subscription needs to be tracked, making the logic and implementation more complex. This subscription logic simply seems to go against the broadcasting semantics of signals. As a final nail in the coffin for this idea, the hashing of the signal name to partition id could be a problem when we want [to dynamically scale the number of partitions](https://github.com/camunda/zeebe/issues/4405). We greatly prefer the simpler solution proposed at the start.

</p>
</details>

## Task breakdown

The task breakdown was done in:
- https://github.com/camunda/zeebe/issues/10851

The topic is split up into several stages.

```[tasklist]
### Tasks
- [x] #11194
- [x] #11918
- [ ] #13241
``` 

## Open questions

- ~~should `signalName` only allowed to be a static value or can it also be a FEEL expression?~~ Signal Name can be static **OR** a FEEL expression
- ~~should signal throw event raise an incident while stage 3 has not yet been implemented?~~ Processes with unsupported elements are rejected

## Links

- [Product Hub Issue](https://github.com/camunda/product-hub/issues/142) (internal)
- [C7 Docs](https://docs.camunda.org/manual/latest/reference/bpmn20/events/signal-events/)

## Special thanks
- To @NPDeehan and @Nlea for their awesome help with the concept process model
- To @skayliu for preparing an overview of the technical task breakdown

**Comment by skayliu** [comment]
Hi, @saig0 @korthout 

In C7 when deployed a process with three top-level start events  in C7, we got three records in the table `act_ru_event_subscr`, but in C8, we got these records separate in there place.

-  `MessageStartEventSubscription` for `Message Start Event` which is already handled by `MessageStartEventSubscriptionManager`
-  `SignalSubscription` for `Signal Start Event` 
-  `ConditionalSubscription` for `Conditional Start Event`, upcoming @lzgabel #11217 



I propose rename the class `MessageStartEventSubscriptionManager` to `StartEventSubscriptionManager`  and move the package of `SubscriptionCommandSender.java` from ` io.camunda.zeebe.engine.processing.message.command` to `package io.camunda.zeebe.engine.processing.command `
for handle the upcoming `Signal` and `Conditioanl` top-level start events.

So, What do you think❓ 


![Process_0kddq72](https://user-images.githubusercontent.com/26370117/206337116-4f4a1ebb-0200-4ba0-9006-61f68f05809d.png)

**Comment by korthout** [comment]
>I propose rename the class `MessageStartEventSubscriptionManager` to `StartEventSubscriptionManager`

This generalization makes sense to me. We should **not merge** the logic to subscribe to Timer start events, as these should only be created on the Deployment Partition (1). We currently have no way to deterministically distribute these timers over multiple partitions, so they are combined on the same partition.

>and move the package of `SubscriptionCommandSender.java` from `io.camunda.zeebe.engine.processing.message.command` to package `io.camunda.zeebe.engine.processing.command`

I think you do not need the `SubscriptionCommandSender` for Signal Events, and also not for Conditional Events. The `SubscriptionCommandSender` is used to communicate specifically between the partition that has the `ProcessMessageSubscription` (the partition where the process instance exists) and the partition that has the `MessageSubscription` (the partition where the message is expected to be published on). 

This is necessary because we don't want to open the message subscription on every partition (that would not be scalable). Instead, the gateway hashes the correlation key of a message to a partition id. That is the partition where the message will be published on. The process instance's partition does the same when it wants to subscribe to a message. It then opens the message subscription on the partition it expects that the message will be published on. To open the message subscription it uses the SubscriptionCommandSender to send the `Open MessageSubscription` command to the other partition. Does that make sense?

Signals will need to be distributed in [stage 2](https://github.com/camunda/zeebe/issues/10851), but special Signal Subscriptions do not have to be opened on other partitions in this same manner. All we need to do is open the SignalSubscription when the Deployment is Created or Distributed to other partitions.

**Comment by madmax25** [comment]
@skayliu to the open question about signalName being static or variable, the signalName is variable in Camunda 7 (backward compatibility) and this flexibility is useful. 

Example use case is creating a signal event listener for multiple running instances that listen for a signalName (dynamically created expression) created from data associated to these running process instances. This allows for more targeted/specific signals that may still need to signal multiple instances.

You may think equivalence can be achieved by sending a variable and having the signal event check the variable conditions to determine if it should proceed or not (conditional if.) I would agree that would suffice however may cause a hiccup on the interrupting signal event, as interruption would probably happen first before the condition check.

**Comment by korthout** [comment]
Thanks for clarifying, @madmax25. That's good to know.

I initially took it out of scope because I wanted to reduce the scope and break it into manageable pieces. However, the first contributions we've received are pointing towards immediately supporting it. If we do happen to keep it out of scope for now, then I'm sure we'll add it shortly after the main parts of this epic are implemented. Your input helps us prioritize it. ❤️

**Comment by abbasadel** [comment]
Blocked by #11456

**Comment by korthout** [comment]
With all three sub epics completed, we can close this issue. Starting with `8.3.0-alpha5`, Zeebe supports Signal Events 🚀 

Thanks again to everyone involved ❤️
